### PR TITLE
Do not set link-layer data via SetInstanceInfo 

### PR DESCRIPTION
**Pull request by manadart**
MAAS was the only provider setting link-layer data via `SetInstanceInfo`, also making it the only provider using the old `SetParentLinkLayerDevicesBeforeTheirChildren` method.

This was causing issues whereby provider IDs were inserted into the global collection, but the devices were set without provider IDs (see deleted code in this patch for specific TODOs regarding this).

In turn the instance-poller was unable to set these IDs, because seeing them in the global collection indicated that they were already in use.

This patch stops `SetInstanceInfo` from setting link-layer data and removes `SetParentLinkLayerDevicesBeforeTheirChildren`. Tests using this method now use `SetLinkLayerDevices` directly. That method is deprecated and has had its validation logic and associated tests removed to ease testing - it is only used in tests.

Work to follow will prevent MAAS from sending link-layer data with `StartInstanceResult` and will probably remove the members from the type completely.

## QA steps

- Bootstrap to MAAS and deploy something (tested on MAAS 2.3 and 2.8).
- Check that there are no logs about not assigning provider IDs to NICs.
- Check that the `linklayerdevices` collection shows documents with provider IDs.

## Documentation changes

None.

## Bug reference

https://bugs.launchpad.net/juju/+bug/1897330

### CODE REVIEW

**Comment by manadart** [comment]
[Do not set link-layer via SetInstanceInfo](https://trello.com/c/AsYRkgl5/3011-do-not-set-link-layer-via-setinstanceinfo)

File: state/machine_linklayerdevices.go [hunk 1]
```diff
@@ -293,75 +293,13 @@ func (m *Machine) prepareOneSetLinkLayerDeviceArgs(
 ) (_ *linkLayerDeviceDoc, err error) {
```
Comment (achilleasa) [inline state/machine_linklayerdevices.go:293]: With this removal the method body looks trivial. Assuming it is not called from different call-sites, should we inline it in the future?

File: state/machine_linklayerdevices.go [hunk 2]
```diff
@@ -489,15 +427,6 @@ func (m *Machine) setDevicesFromDocsOps(newDocs []linkLayerDeviceDoc) ([]txn.Op,
 
 func (m *Machine) insertLinkLayerDeviceOps(newDoc *linkLayerDeviceDoc) ([]txn.Op, error) {
 	var ops []txn.Op
-	if newDoc.ParentName != "" {
-		newParentDocID, err := m.parentDocIDFromDeviceDoc(newDoc)
-		if err != nil {
-			return nil, errors.Trace(err)
-		}
-		if newParentDocID != "" {
-			ops = append(ops, assertLinkLayerDeviceExistsOp(newParentDocID))
```
Comment (achilleasa) [inline state/machine_linklayerdevices.go:498]: Does the facade code guarantee that devices are inserted in the correct order (and/or uses such an assertion in the generated txn op list)? If not, how do we ensure that we are not missing any parents?

Comment (manadart) [reply state/machine_linklayerdevices.go:498]: The facade code does not call though this path, but you have turned up that we do not actually have this check.

We *do* note observed parent devices up-front in order to ensure that devices referred to by children are not removed. I'll add another ticket for this part.

File: state/machine_linklayerdevices.go (hunk 1 above)
Comment (manadart) [reply state/machine_linklayerdevices.go:293]: I'll be deleting it once the tests use an alternate method for priming link-layer data.

File: state/machine_linklayerdevices.go [hunk 3]
```diff
@@ -198,9 +198,9 @@ func (m *Machine) AddLinkLayerDeviceOps(
 // - Machine is no longer alive or is missing;
 // - Model no longer alive;
 // - errors.NotValidError, when any of the fields in args contain invalid values;
-// - ErrProviderIDNotUnique, when one or more specified ProviderIDs are not unique;
-// Setting new parent devices must be done in a separate call than setting their
-// children on the same machine.
+//
+// Deprecated: (manadart 2020-10-12) This method is only used by tests and is in
+// the process of removal. Do not add new usages of it.
 func (m *Machine) SetLinkLayerDevices(devicesArgs ...LinkLayerDeviceArgs) (err error) {
```
[the remark this answers is not in the document]
Comment (manadart) [reply state/machine_linklayerdevices.go:204]: This is done.

**Comment by achilleasa** [comment]
The extra commit looks fine but there seems to be an issue with the tests.

**Review by achilleasa — APPROVED** [review]

**Comment by manadart** [comment]
$$merge$$

**Pull request merged**